### PR TITLE
8281116: [lworld] Adding Preload attribute support

### PR DESCRIPTION
**Pull request by fparain**
Please review this changeset adding support for the Preload attribute.
It includes a logging feature to track and diagnose those optimistic class loading.

Tested with Mach5 tier1.

Thank you,

Fred

---------
### Progress
- [x] Change must not contain extraneous whitespace

### Issue
 * [JDK-8281116](https://bugs.openjdk.java.net/browse/JDK-8281116): [lworld] Adding Preload attribute support

### Reviewers
 * [Harold Seigel](https://openjdk.java.net/census#hseigel) (@hseigel - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/valhalla pull/623/head:pull/623` \
`$ git checkout pull/623`

Update a local copy of the PR: \
`$ git checkout pull/623` \
`$ git pull https://git.openjdk.java.net/valhalla pull/623/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 623`

View PR using the GUI difftool: \
`$ git pr show -t 623`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/valhalla/pull/623.diff">https://git.openjdk.java.net/valhalla/pull/623.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/oops/instanceKlass.hpp [hunk 1]
```diff
@@ -285,7 +285,8 @@ class InstanceKlass: public Klass {
     _misc_invalid_inline_super                = 1 << 19, // invalid super type for an inline type
     _misc_invalid_identity_super              = 1 << 20, // invalid super type for an identity type
     _misc_has_injected_identityObject         = 1 << 21, // IdentityObject has been injected by the JVM
-    _misc_has_injected_primitiveObject        = 1 << 22  // PrimitiveObject has been injected by the JVM
+    _misc_has_injected_primitiveObject        = 1 << 22, // PrimitiveObject has been injected by the JVM
+    _misc_has_preload_attribute               = 1 << 23  // class has a Preload attribute
   };
```
Comment (hseigel) [inline src/hotspot/share/oops/instanceKlass.hpp:291]: Is the _misc_has_preload_attribute flag needed, or can the code just check if _preload_classes equals Null?

Comment (fparain) [reply src/hotspot/share/oops/instanceKlass.hpp:291]: Strictly speaking, it is not needed. At some point, I've considered putting the _preload_classes pointer into the optional part of InstanceKlass, which would have required this flag. But finally it is declared as a regular field, so this flag can be removed.

File: src/hotspot/share/classfile/classFileParser.cpp [hunk 2]
```diff
@@ -5884,6 +5945,7 @@ ClassFileParser::ClassFileParser(ClassFileStream* stream,
   _nest_members(NULL),
   _nest_host(0),
   _permitted_subclasses(NULL),
+  _preload_classes(NULL),
   _record_components(NULL),
   _temp_local_interfaces(NULL),
   _local_interfaces(NULL),
```
Comment (hseigel) [inline src/hotspot/share/classfile/classFileParser.cpp:5951]: Do you need to free _preload_classes in ClassFileParser::~ClassFileParser() and set _preload_classes to NULL in ClassFileParser::clear_class_metadata(), also, assert that it is NULL in ClassFileParser::fill_instance_klass() ?

File: test/hotspot/jtreg/runtime/valhalla/inlinetypes/ValuePreloadTest.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    public static void main(String[] args) throws Exception {
+        ProcessBuilder pb = exec("-Xlog:class+preload","ValuePreloadClient0");
+        checkFor(pb, "[info][class,preload] Preloading class PreloadValue0 during linking of class ValuePreloadClient0 because of its Preload attribute");
+
+        pb = exec("-Xlog:class+preload","ValuePreloadClient1");
+        checkFor(pb, "[warning][class,preload] Preloading of class PreloadValue1 during linking of class ValuePreloadClient1 (Preload attribute) failed");
+    }
+}
```
[the remark this answers is not in the document]
Comment (fparain) [reply test/hotspot/jtreg/runtime/valhalla/inlinetypes/ValuePreloadTest.java:61]: Fixed

File: src/hotspot/share/classfile/classFileParser.cpp (hunk 2 above)
Comment (fparain) [reply src/hotspot/share/classfile/classFileParser.cpp:5949]: Fixed

**Comment by fparain** [comment]
Harold,

Thank you for your feedback. Those issues should be fixed with the last commit.

Fred

**Review by hseigel — APPROVED** [review]
The changes look good!
Thanks, Harold

**Pull request closed without merging**